apps/shop/tests.py

```diff
@@ -6,34 +9,168 @@
+        item: Item = G(Item, available=False, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_saldo_corresponds_with_purchase_amount(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+        self.assertEqual(self.user.saldo, item.price)
+
+        self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(self.user.saldo, 0)
+
+    def test_purchase_requires_an_item_quantity(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id}])
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_purchase_quantity(self):
+        quantity = 3
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price * quantity)
+
+        response = self._perform_purchase(
+            [{"object_id": item.id, "quantity": quantity}]
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(self.user.saldo, 0)
+
+    def test_purchase_quantity_requires_saldo_for_all(self):
+        price = 100
+        quantity = 3
+        item: Item = G(Item, available=True, price=price)
+        self._add_saldo_to_user(price)
+
+        response = self._perform_purchase(
+            [{"object_id": item.id, "quantity": quantity}]
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
+        self.assertEqual(response.json().get("detail"), "Insufficient funds")
+
+    def test_purchase_multiple_items_at_once(self):
+        item1_quantity = 1
+        item2_quantity = 2
+        item1: Item = G(Item, available=True, price=100)
+        item2: Item = G(Item, available=True, price=300)
+
+        self._add_saldo_to_user(
+            item1.price * item1_quantity + item2.price * item2_quantity
+        )
+
+        response = self._perform_purchase(
+            [
+                {"object_id": item1.id, "quantity": item1_quantity},
+                {"object_id": item2.id, "quantity": item2_quantity},
+            ]
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(self.user.saldo, 0)
+
+    def test_purchase_triggers_receipt(self):
+        item: Item = G(Item, available=True, price=100)
+
+        self.assertEqual(len(mail.outbox), 0)
+        self._add_saldo_to_user(item.price)
+
+        self.assertEqual(len(mail.outbox), 1)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(len(mail.outbox), 2)
```

Comment:
I think it would be beneficial to also test that the `to`-address is correct as well, and perhaps that the correct commitee is CC'd?

Reply:
Maybe, but I did not find it necessary.
The email things should not be in the domain of the `shop` app, and I don't think anyone is supposed to be CC-ed for a purchase in the shop